shared/actions/fs/index.tsx

```diff
@@ -963,42 +963,59 @@ const checkKbfsServerReachabilityIfNeeded = (state, action: ConfigGen.OsNetworkS
 const onFSOnlineStatusChanged = (state, action: EngineGen.Keybase1NotifyFSFSOnlineStatusChangedPayload) =>
   FsGen.createKbfsDaemonOnlineStatusChanged({online: action.payload.params.online})
 
-const onFSOverallSyncSyncStatusChanged = (
+const onNotifyFSOverallSyncSyncStatusChanged = (
   state,
   action: EngineGen.Keybase1NotifyFSFSOverallSyncStatusChangedPayload
-) =>
-  FsGen.createOverallSyncStatusChanged({
-    outOfSpace: action.payload.params.status.outOfSyncSpace,
-    progress: Constants.makeSyncingFoldersProgress(action.payload.params.status.prefetchProgress),
-  })
-
-const notifyDiskSpaceStatus = (diskSpaceStatus: Types.DiskSpaceStatus) => {
-  switch (diskSpaceStatus) {
-    case Types.DiskSpaceStatus.Error:
-      NotifyPopup('Sync Error', {
-        body: 'You are out of disk space. Some folders could not be synced.',
-        sound: true,
-      })
-      break
-    case Types.DiskSpaceStatus.Warning:
-      NotifyPopup('Disk Space Low', {body: 'You have less than 1 GB of storage space left.'})
-      break
-    case Types.DiskSpaceStatus.Ok:
-      break
-    default:
-      Flow.ifFlowComplainsAboutThisFunctionYouHaventHandledAllCasesInASwitch(diskSpaceStatus)
-  }
-}
-
-let prevOutOfSpace = false
-const updateMenubarIconOnStuckSync = (state, action) => {
-  const outOfSpace = action.payload.params.status.outOfSyncSpace
-  if (outOfSpace !== prevOutOfSpace) {
-    prevOutOfSpace = outOfSpace
-    // TODO once go side sends info: low on space warning
-    notifyDiskSpaceStatus(outOfSpace ? Types.DiskSpaceStatus.Error : Types.DiskSpaceStatus.Ok)
-    return NotificationsGen.createBadgeApp({key: 'outOfSpace', on: outOfSpace})
+) => {
+  const diskSpaceStatus = action.payload.params.status.outOfSyncSpace
+    ? Types.DiskSpaceStatus.Error
+    : action.payload.params.status.localDiskBytesAvailable <
+      state.fs.settings.spaceAvailableNotificationThreshold
+    ? Types.DiskSpaceStatus.Warning
+    : Types.DiskSpaceStatus.Ok
+  // We need to type this separately since otherwise we can't concat to it.
+  let actions: Array<
+    | NotificationsGen.BadgeAppPayload
+    | FsGen.OverallSyncStatusChangedPayload
+    | FsGen.ShowHideDiskSpaceBannerPayload
+  > = [
+    FsGen.createOverallSyncStatusChanged({
+      diskSpaceStatus,
+      progress: Constants.makeSyncingFoldersProgress(action.payload.params.status.prefetchProgress),
+    }),
+  ]
+  // Only notify about the disk space status if it has changed.
```

Comment:
Do actions guaranteedly get run before reducers?

Reply:
It shouldn't matter since `state` is immutable point-in-time.

Reply:
If the reducer runs first, then the check of whether the status has changed will be incorrect (think it hasn't changed even though it has).  Or is there not a reducer running on this trigger?

Reply:
OK I understand your question now. I believe Sagas are guaranteed to be called with the pre-reduced state.

Reply:
Actually, I'm googling around a bit now and it seems you might be correct. But we've built functionality with this assumption before, and now I'm wondering how that functionality works.

Reply:
@jakob223 I further thought about it, and I think it's actually fine here. Since the action that triggers this handler doesn't actually change any state in the reducer. The action that changes state is the one triggered on the line above. But that one is independent of the state variable in this saga handler.